lib/factory_bot/definition.rb

```diff
@@ -114,14 +114,28 @@ def trait_by_name(name)
       trait_for(name) || Internal.trait_by_name(name)
     end
 
+    # This is a hot path when using traits so:
+    # - Memoize the defined_traits by name so we can lookup in a hash instead of array
+    # - Store both string/symbol names to avoid needing to cast on lookup
+    # - Use plain Hash since HashWithIndifferentAccess incurs significant overhead
     def trait_for(name)
-      defined_traits.detect { |trait| trait.name == name.to_s }
+      @defined_traits_by_name ||= defined_traits.each_with_object({}) do |trait, memo|
+        # This method was previously using `detect` which means if there were
```

Comment:
I think this comment makes more sense in the body of the commit message. The specific spec and line number are likely to change, which would make the comment inaccurate.

Reply:
Good call out. My intention was to explicitly call out behavior that’s a bit implicit but I may try adding a test for this purpose instead.

Reply:
I added tests to more explicitly test this behavior. It ensures that traits defined multiple times in the same factory will use the first one and you're able to redefine the trait in child factories. For the former (defining a trait twice in same factory) seems like it's something that should raise an exception although that would be a breaking change and definitely outside of the scope of this PR.